### DISCUSSION TITLE
What are nesting factories used for?

### DISCUSSION THREAD

**Discussion by weltkante** [microsoft/vs-threading#816] [Q&A] [answered]
I'm currently studying the codebase to evaluate how useful a port of the deadlock resolution to async frameworks on different runtimes would be and am a bit confused on the purpose of `nestingFactories`.

From what I can tell they contain a set of `JoinableTaskFactory` of its ambient tasks. This is [calculated](https://github.com/microsoft/vs-threading/blob/d5d6130ddd5bffb762c1bf13bf78dc356b8993e4/src/Microsoft.VisualStudio.Threading/JoinableTaskFactory.cs#L1019) whenever a `JoinableTask` is launched.

The publically visible call sites to code which consumes `nestingFactories` are `SynchronizationContext.Post` and `SwitchToMainThread`. The usage of `nestingFactories` seems to be that it [will post](https://github.com/microsoft/vs-threading/blob/e5907f1fa83efe71b4af387899d7551668315797/src/Microsoft.VisualStudio.Threading/JoinableTask.cs#L768) to any `UnderlyingSynchronizationContext` in the hierarchy of ambient tasks.

So whats the point of all this effort? It is optimized to not allocate in the common case of only having a single `JoinableTaskFactory`, so obviously there is some issue with having multiple TaskFactories it tries to solve? There are two scenarios I can think of that lead to having multiple factories:
- Creating multiple instances of `JoinableTaskFactory` for the same `JoinableTaskContext` (including `JoinableTaskCollection`)
- Creating multiple instances from distinct `JoinableTaskContext` (different UI threads?)

Since `UnderlyingSynchronizationContext` forwards to the `JoinableTaskContext` the first case does not really solve any issues does it? It just posts multiple times to the same SC, so it actually is an unnecessary performance overhead if anything, isn't it? I'd have expected the set of nesting `JoinableTaskContext` to be stored, not the set of nesting `JoinableTaskFactory`.

If the factories are for different UI threads then I'm wondering if this code is even correct in the first place, why would it post to every underlying thread when targeting a specific JoinableTask? Doesn't this basically mean code can run on arbitrary threads which happen to be in the chain launching a `JoinableTask`?

So, any idea what the nesting factories code is there for? Any scenario it actually helps with?

**Comment by AArnott** [comment]
FYI @sharwell actually *did* port this to Java several years ago. At least, I think he did, but now I can't find anything on [an Internet search](https://www.bing.com/search?q=java+joinabletaskfactory&cvid=0d0177b88c3a40689cfeb1d0957ab948&aqs=edge..69i57.5231j0j1&pglt=43&FORM=ANNAB1&PC=U531) to back that up.

**Reply by weltkante** [reply]
I happened to run across it, in case anyone else is interested: https://github.com/tunnelvisionlabs/java-threading

(Google finds it while Bing doesn't - seems Bing needs to improve its github indexing)

**Comment by AArnott** [comment, marked as answer]
> Creating multiple instances from distinct JoinableTaskContext (different UI threads?)

This is _not_ the reason. Multiple instances of `JoinableTaskContext` *never* talk to each other. If you have two threads in the process that each have a single-threaded `SynchronizationContext` applied, then conceivably two `JoinableTaskContext` instances may apply in that process, but they will not communicate and deadlocks may result if any code needs *both* threads to complete.

>  I'd have expected the set of nesting JoinableTaskContext to be stored, not the set of nesting JoinableTaskFactory.

There is no such thing as nested `JoinableTaskContext`. The contract is to have exactly one for a designated thread.

You *may* create multiple `JoinableTaskFactory` instances, and those factories vary not only in which `JoinableTaskCollection` they may be associated with but also (and importantly for the point of `nestingFactories`) in *priority* of how they get to the UI thread.
Consider WPF, which has a `Dispatcher` by which you can schedule work with a variety of priorities back to the dispatcher thread. You can represent each of these priorities with a different `JoinableTaskFactory` instance using a derived type of that class that overrides the method responsible for calling `SynchronizationContext.Post` to instead call `Dispatcher.InvokeAsync` with some preset priority.
When a JoinableTask requires the main thread, we want to get it there with the highest priority found within itself and any parent JoinableTasks.

You can see tests related to nested factories and this priorities scenario starting with this one (and others nearby in that file):

https://github.com/microsoft/vs-threading/blob/f5782224701062eb95d65662c012b82e02dd5e89/test/Microsoft.VisualStudio.Threading.Tests/JoinableTaskTests.cs#L2686-L2705

**Reply by weltkante** [reply]
Thanks, the explanation is very much appreciated.

I now see that I missed `DispatcherJoinableTaskFactory` when analyzing call chains because it was disabled via `#if` in the build I was looking at. Also was relying too much on VS for navigating the source, maybe should have done a text search to find comments/tests. Embarrassing mistakes.
